Liberty olo with rh sso (#3390)
OpenLiberty/blogs: posts/2024-08-20-liberty-olo-rh-sso.adoc

RUN configure.sh

----

* Ensure that you set the `ENV SEC_TLS_TRUSTDEFAULTCERTS` and `ENV SEC_IMPORT_K8S_CERTS` environment variables to true so you can trust all the certificates from within the cluster.
* By specifying `ARG SEC_SSO_PROVIDERS="oidc"`, you tell the configuration that the SSO provider is OIDC.

You can find out more about the available configuration options in the Open Liberty container image link:https://github.com/OpenLiberty/ci.docker/blob/main/SECURITY.md#single-sign-on-configuration[Open Liberty container image SSO documentation].

Now, we can move on to the next step, installing and configuring the Red Hat OpenShift cluster to deploy this application to.

[#installRHSSO]
== Install and configure the RH-SSO Operator in the Red Hat OpenShift cluster

Complete the following steps to set up the Red Hat OpenShift cluster so that you can use RH-SSO.

. To install RH-SSO in the `rh-sso` namespace, follow the instructions that are provided in the link:https://access.redhat.com/documentation/en-us/red_hat_single_sign-on/7.6/html/server_installation_and_configuration_guide/operator#doc-wrapper[Red Hat Single Sign-On Operator documentation].

> . Create and log in to a KeyCloak instance.

Reviewer: I think this step needs more detail

Author: I can add more details by adding screen-shots. Will that help?
These are for installing the operator
<img width="896" alt="Screenshot 2024-08-15 at 4 39 37 PM" src="https://github.com/user-attachments/assets/ac8b8fa8-b3f0-43a1-ad8c-63db68523e28">

<img width="1854" alt="Screenshot 2024-08-15 at 4 41 49 PM" src="https://github.com/user-attachments/assets/bfa8e08c-3c19-4b03-b85c-58c57b808651">

Reviewer: yes- i think screen shots of the console and a description of the steps would be helpful. Alternatively, if this is already documented in detail by redhat, we can provide a link to that specific section of their doc.

+

> After you install the RH-SSO Operator, create a KeyCloak instance that uses the default values. You can then access KeyCloak by looking at the routes. The route is in the `\https://keycloak-rh-sso.apps.<cluster_name>` format. You use this URL to log in to the KeyCloak instance. The credentials for logging in are in the `credential-example-keycloak` secret in the `rh-sso` namespace. Get the secret password for the `admin` username in this secret, then use this username and password to log in on to KeyCloak.

Reviewer: > After you install the RH-SSO Operator, create a KeyCloak instance that uses the default values.
How? Is it the same process described in RH docs [11.3.2. Creating a Keycloak custom resource on OpenShift](https://docs.redhat.com/en/documentation/red_hat_single_sign-on/7.6/html/server_installation_and_configuration_guide/operator#creating_a_keycloak_custom_resource_on_openshift) ? If so, we should link to it. I realize this section relies heavily on RH docs but we should make it as easy as we can for user to find the right info.

> You can then access KeyCloak by looking at the routes. 
Here too- how do they find the routes? 

> The route is in the `\https://keycloak-rh-sso.apps.<cluster_name>` format.

When I look at the "Keycloak custom resource Status" example in the RH doc, the route is not in the format we suggest here.

Author: This screenshot shows how to create the KeyCloak instance.

<img width="1333" alt="Screenshot 2024-08-15 at 4 46 24 PM" src="https://github.com/user-attachments/assets/58809c10-d12d-411c-aba7-74bd3aabb974">

And this screenshot here shows the routes for the rh-sso namespace in my cluster.

<img width="1710" alt="Screenshot 2024-08-15 at 5 05 53 PM" src="https://github.com/user-attachments/assets/dc6adecc-28c2-441f-8915-c34310bb1f60">

As you can see - it is in the format mentioned above:

`https://keycloak-rh-sso.apps.rhaque-gsm.cp.fyre.ibm.com`

Reviewer: ok- either a link to the red hat doc section or screen shots and description would be helpful, and save the user time if they are not already familiar with RH SSO

> . Create a realm named `sso-realm`. Use the following URL to access this realm:

Reviewer: > Create a realm named `sso-realm`. 

Does this map to [Creating a realm custom resource](https://docs.redhat.com/en/documentation/red_hat_single_sign-on/7.6/html/server_installation_and_configuration_guide/operator#realm-cr) in the RH doc?

Author: Yes - this document shows how to create a Realm in RH-SSO.

> . Create a realm named `sso-realm`. Use the following URL to access this realm:

Reviewer: ```suggestion
. link:https://docs.redhat.com/en/documentation/red_hat_single_sign-on/7.6/html/server_installation_and_configuration_guide/operator#realm-cr[Create a realm] named `sso-realm`. Use the following URL to access this realm:
```

+
[.img_border_light]
image::img/blog/rh_testuser1_roles.png[Roles for testuser1,width=50%,align="center"]

. Test the user you created.
+
Go to the `\https://keycloak-rh-sso.apps.<cluster-name>/auth/admin/Sso-realm/console/` URL and log in by using the `testuser1`/`testpasswd1` username you created in the previous steps. After you log in, in the **General** section, you can see the endpoints.
. Click the link for the **OpenID Endpoint Configuration**, which points you to the following URL:
+
`\https://keycloak-rh-sso.apps.<cluster-name>/auth/realms/sso-realm/.well-known/openid-configuration`.
+
You need this URL as the `discoveryEndpoint` for the client registration later on.

[#createSecret]
== Create the olapp-sso secret

Next, we create a secret for the Open Liberty application. The key name for this secret is in the `<app-name>-olapp-sso` format. You must use the same `<app-name>` as the one you use to deploy the application with OLO. For example, in the application that is used in this post, the `<app-name>` from the YAML file is `guide-social-media-login`, so the secret name is `guide-social-media-login-olapp-sso`.

> . Create a project in your cluster called `gsm-test`.

Reviewer: > Create a project in your cluster called `gsm-test`.

How? is this documented on the RH side somewhere that we can point to?

Author: https://docs.openshift.com/container-platform/4.16/applications/projects/working-with-projects.html documents how to create a project.

> . Create a project in your cluster called `gsm-test`.

Reviewer: ```suggestion
. link:https://docs.openshift.com/container-platform/4.16/applications/projects/working-with-projects.html[Create a project] in your cluster called `gsm-test`.
```

After installing the Open Liberty Operator, use the following YAML file to deploy the `guide-social-media-login` application that you previously created.

[source]
----
apiVersion: apps.openliberty.io/v1
kind: OpenLibertyApplication
metadata:
  name: guide-social-media-login
  namespace: gsm-test
spec:
  sso:
    oidc:
      - discoveryEndpoint:
          <https://keycloak-rh-sso.apps.<cluster-name>/auth/realms/sso-realm/.well-known/openid-configuration>
  service:
    port: 9443
  applicationImage:
    <image location of the app>
  expose: true
  manageTLS: true
  replicas: 1
  applicationName: guide-sm-login
  pullPolicy: Always
  pullSecret: <secret_to_pull_image>

----

* The name of the application is `guide-social-media-login`, the same name that was used when you created the previous secret.
* The `oidc: discoveryEndpoint` must point to the OpenID endpoint configuration that you set when you configured the RH-SSO Operator.
* The `applicationImage` value must point to your image location where you placed your application image, in a container registry like DockerHub or Artifactory.

> * The `pullSecret` value must be set to be able to access the container registry.

Reviewer: > The `pullSecret` value must be set to be able to access the container registry.

what should it be set to- is it the `guide-social-media-login-olapp-sso` secret they created in the previous step?

Author: No, this secret is different.
After you upload your container image (containing your application) to a container registry, besides pointing to the url that has the image, you also need to provide it the credentials to be able to access that registry.

In my case, I was using Artifactory to push the image to, and I have my pull-secret pointing to the credentials that I use to log into Artifactory.

This part is more generic, because we can use any type of container registry (such as dockerhub etc). But basically after you point to the image, you also have to provide the credentials that OCP will need to pull down that image.

> * The `pullSecret` value must be set to be able to access the container registry.

Reviewer: ```suggestion
* The `pullSecret` value must be set to the name of the secret that contains the credentials for your chosen container registry.
```
